### PR TITLE
Add PySpark unit testing

### PR DESCRIPTION
**Pull request by henryre**
On the way to adding integration/complex testing, I found a good way to add PySpark tests. We don't want to run them with everything else all the time because we need Java and the PySpark install is huge. We do however want to run them in Travis.

You can also see that I'm marking things with `pytest.mark.complex`. In a future diff, we'll add integration/complex system testing support by just running `pytest.mark.complex` on master.

It's important to note that single node Spark tests won't catch everything. But it's the best we can do with current CI setup.

* Add PySpark unit tests, decorated with `pytest.mark.spark`
* Separate Spark tests from other tests in pytest invocation in tox.ini
* Update Travis to install JDK8 so that we can run Spark tests
* Add documentation on PySpark testing

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -85,6 +85,17 @@ Note that we use PEP 484 type hints, so parameter types should be removed from t
 }
 ```
 
+### PySpark tests
+PySpark tests are invoked separately from the rest since they require
+installing Java and the large PySpark package.
+They are executed on Travis, but not by default for a local `tox` command.
```
Comment (paroma) [inline CONTRIBUTING.md]: repeating discussion we had earlier, TBD a way to not install packages every time a PR is created/updates are pushed to Travis

File: requirements.txt [hunk 2]
```diff
@@ -40,10 +40,9 @@ tensorboardX>=1.6
 # Scale-up/out
 dask[complete]>=1.2.1
 
-# Note: you may want to remove PySpark from the 
-# required installs. Installing a new version may
-# overwrite your existing system install.
-pyspark>=2.4.3
+# Note: we don't include PySpark in the required installs.
+# Installing a new version may overwrite your existing system install.
+# pyspark>=2.4.3
```
Comment (paroma) [inline requirements.txt]: we don't want  == some version that works for the tests?

File: CONTRIBUTING.md (hunk 1 above)
Comment (henryre) [reply CONTRIBUTING.md]: You mean the `tox` setup steps on Travis are taking too long?

File: requirements.txt (hunk 2 above)
Comment (henryre) [reply requirements.txt]: yeah I could make a requirements-spark.txt file to pin it just for tests

File: CONTRIBUTING.md (hunk 1 above)
Comment (henryre) [reply CONTRIBUTING.md]: yeah...seems like caching pip (which we already do) is the best we can do. i can keep digging.

**Comment by paroma** [comment]
In general, the `lf_applier` and `map` tests are repeated per data type within the applier file and then again with Spark, is there a way to get around this somehow?

**Comment by henryre** [comment]
@paroma good point. kind of trading off readability for maintainability (maybe) here. as in, programmatically generated test cases are more difficult to reason with manually. any ideas coming to mind for a clean up?

File: CONTRIBUTING.md (hunk 1 above)
Comment (henryre) [reply CONTRIBUTING.md]: Opened a task in Asana, pushing on this now.

**Pull request merged**